cmd/endpoint.go

```diff
@@ -0,0 +1,145 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package cmd
+
+import (
+	"fmt"
+	"net/url"
+
+	"github.com/minio/minio-go/pkg/set"
+)
+
+// IsValidDistribution - checks whether given count is a valid distribution for erasure coding.
+func IsValidDistribution(count int) bool {
+	return (count >= 4 && count <= 16 && count%2 == 0)
+}
+
+// IsEmptyPath - check whether given path is not empty.
+func IsEmptyPath(path string) bool {
+	return path == "" || path == "." || path == "/" || path == `\`
+}
+
+// EndpointType - enum for endpoint type.
+type EndpointType int
+
+const (
+	// UnknownEndpointType - unknown endpoint type enum.
+	UnknownEndpointType EndpointType = 0
+
+	// FSEndpointType - FS endpoint type enum.
+	FSEndpointType = 1
+
+	// XLEndpointType - XL endpoint type enum.
+	XLEndpointType = 2
+)
+
+// Endpoint - any type of endpoint.
+type Endpoint struct {
+	Index      int
+	Value      string
+	IsDisabled bool
+	URL        *url.URL
+}
+
+// Type - returns type of endpoint.
+func (endpoint Endpoint) Type() EndpointType {
+	if IsEmptyPath(endpoint.Value) {
+		return UnknownEndpointType
+	}
+
+	if endpoint.URL != nil {
+		return XLEndpointType
+	}
+
+	return FSEndpointType
+}
+
+// NewEndpoint - returns new endpoint based on given arguments.
+func NewEndpoint(arg string, index int) (Endpoint, error) {
+	if IsEmptyPath(arg) {
+		return Endpoint{}, fmt.Errorf("Empty or root endpoint is not supported")
+	}
+
+	u, err := url.Parse(arg)
+	if err == nil && u.Host != "" {
+		// May be XL type endpoint arg.
+		host, port, err := splitHostPort(u.Host)
+		if !(err == nil && host != "" && (u.Scheme == "http" || u.Scheme == "https") &&
+			u.Opaque == "" && u.ForceQuery == false && u.RawQuery == "" && u.Fragment == "") {
+			return Endpoint{}, fmt.Errorf("Unknown endpoint format")
+		}
+
+		if IsEmptyPath(u.Path) {
+			return Endpoint{}, fmt.Errorf("Empty or root path is not supported in URL endpoint")
+		}
+
+		u.Host = host + ":" + port
+		arg = u.String()
+	} else {
+		u = nil
+	}
+
+	return Endpoint{
+		Index: index,
+		Value: arg,
+		URL:   u,
+	}, nil
+}
+
+// EndpointList - endpoint list hashed with endpoint's value
+type EndpointList map[string]Endpoint
+
+// NewEndpointList - returns new endpoint list based on input args.
+func NewEndpointList(args ...string) (endpoints EndpointList, err error) {
+	// Check whether given args contain duplicates.
+	if uniqueArgs := set.CreateStringSet(args...); len(uniqueArgs) != len(args) {
+		return nil, fmt.Errorf("duplicate endpoints found")
+	}
+
+	// Check whether no. of args are valid for XL distribution.
+	if !IsValidDistribution(len(args)) {
```

Comment:
IsValidDistribution() accepts only one argument, which means NewEndpointList works only with XL/Distributed (not obvious), can we make this check outside this function, it will be less confusing.